packages/circuit-ui/components/Dialog/Dialog.tsx

```diff
@@ -0,0 +1,345 @@
+
+      if (!hasNativeDialog && isModal) {
+        (dialogElement.nextSibling as HTMLDivElement).classList.remove(
+          classes['backdrop-visible'],
+        );
+      }
+      onCloseStart?.();
+      // trigger closing of the dialog after animation
+      setTimeout(() => {
+        if (dialogElement.open) {
+          dialogElement.close();
+        }
+      }, animationDuration);
+    }, [hasNativeDialog, animationDuration, isModal, onCloseStart]);
+
+    useClickOutside([dialogRef], () => handleDialogClose(), open && !isModal);
+    useEscapeKey(() => handleDialogClose(), open && !isModal);
+
+    useEffect(() => {
+      const dialogElement = dialogRef.current;
+
+      if (!dialogElement) {
+        return undefined;
+      }
+
+      if (!open) {
+        dialogElement.returnValue = '';
+      }
+      if (open && !dialogElement.open) {
+        if (isModal) {
+          dialogElement.showModal();
+        } else {
+          dialogElement.show();
+        }
+        if (!hasNativeDialog && dialogElement.nextSibling) {
+          // use the polyfill backdrop
+          (dialogElement.nextSibling as HTMLDivElement).classList.add(
+            classes['backdrop-visible'],
+            classes.backdrop,
+          );
+          // intercept and prevent modal closing if preventClose is true
+          (dialogElement.nextSibling as HTMLDivElement).addEventListener(
+            'click',
+            onPolyfillBackdropClick,
+          );
+        }
+      }
+      if (!open && dialogElement.open) {
+        handleDialogClose();
+      }
+
+      return () => {
+        if (dialogElement.open) {
+          dialogElement.close('skipOnClose');
+        }
+        (dialogElement.nextSibling as HTMLDivElement)?.removeEventListener(
+          'click',
+          onPolyfillBackdropClick,
+        );
+      };
+    }, [
+      isModal,
+      open,
+      handleDialogClose,
+      hasNativeDialog,
+      onPolyfillBackdropClick,
+    ]);
+
+    const onDialogClick = (
+      event: ClickEvent<HTMLDialogElement> | ClickEvent<HTMLDivElement>,
+    ) => {
+      // the dialog content covers the whole dialog element
+      // leaving the backdrop element as the only clickable area
+      // that can trigger an onClick event
+      if (event.target === event.currentTarget && !preventClose) {
+        handleDialogClose();
+      }
+    };
+
+    return (
+      <>
+        {/* eslint-disable-next-line  jsx-a11y/no-noninteractive-element-interactions */}
+        <dialog
+          onClick={onDialogClick}
+          className={clsx(classes.base, isModal && classes.modal, className)}
+          ref={applyMultipleRefs(ref, dialogRef)}
+          {...rest}
+        >
+          {isModal && (
+            <CloseButton onClick={handleDialogClose} className={classes.close}>
+              {closeButtonLabel}
+            </CloseButton>
+          )}
+          {open &&
+            (typeof children === 'function'
+              ? children?.({ onCloseEnd })
```

Comment:
Hm, I figured `onCloseStart` and `onCloseEnd` would only be used internally by the higher-level components such as Modal or Popover. Given that the `children` prop will make it into the public API of those components, I would keep the less verbose `onClose` name here. 
```suggestion
              ? children?.({ onClose: onCloseEnd })
```